#### Embedding Techniques

to convert text to vectors

In [1]:
import os
from dotenv import load_dotenv
load_dotenv()

True

In [3]:
os.environ["OPENAI_API_KEY"]=os.getenv('OPENAI_API_KEY')

In [4]:
from langchain_openai import OpenAIEmbeddings
embeddings = OpenAIEmbeddings(model="text-embedding-3-large")

/home/hb/miniconda3/envs/agentic_env/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [5]:
embeddings

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x7fd404c3c560>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x7fd4043a62d0>, model='text-embedding-3-large', dimensions=None, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [6]:
text="This is a tutorial on OPENAI embeddings"
query_result=embeddings.embed_query(text)
query_result

[-0.0041656494140625,
 0.0408935546875,
 -0.019012451171875,
 -0.040313720703125,
 0.0225830078125,
 0.0010929107666015625,
 0.0123291015625,
 0.052154541015625,
 -0.0103759765625,
 0.004917144775390625,
 -0.0171661376953125,
 0.00634765625,
 -0.01020050048828125,
 -0.01165008544921875,
 0.01122283935546875,
 0.0181884765625,
 0.0223541259765625,
 0.04925537109375,
 -0.0180511474609375,
 -0.0203857421875,
 0.0024318695068359375,
 0.006923675537109375,
 -0.0498046875,
 -0.039642333984375,
 0.0262603759765625,
 -0.019805908203125,
 0.01050567626953125,
 0.039031982421875,
 -0.0088348388671875,
 0.0518798828125,
 0.012664794921875,
 -0.01132965087890625,
 0.0029125213623046875,
 -0.00206756591796875,
 -0.0153656005859375,
 0.0190582275390625,
 0.04278564453125,
 0.032989501953125,
 -0.040374755859375,
 0.0238494873046875,
 0.0294036865234375,
 0.0037822723388671875,
 0.0022563934326171875,
 -0.0216064453125,
 -0.0511474609375,
 -0.0273590087890625,
 0.0081787109375,
 -0.0257415771484375,


In [7]:
len(query_result)

3072

The openAI generates embeddings to a fixed length of 3072. We can also control the number of dimensions also.

In [8]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-large", dimensions=1024)

In [8]:
text="This is a tutorial on OPENAI embeddings"
query_result=embeddings.embed_query(text)
query_result

[-0.005755920894443989,
 0.056403547525405884,
 -0.026195039972662926,
 -0.05561518669128418,
 0.031229790300130844,
 0.0015196079621091485,
 0.016976607963442802,
 0.07195573300123215,
 -0.01435172837227583,
 0.006795122288167477,
 -0.02366870455443859,
 0.008770500309765339,
 -0.01409192755818367,
 -0.01612553745508194,
 0.015426764264702797,
 0.02510208636522293,
 0.030799776315689087,
 0.06801393628120422,
 -0.024887079373002052,
 -0.02818385511636734,
 0.0033415695652365685,
 0.009531984105706215,
 -0.06869478523731232,
 -0.054611820727586746,
 0.036282457411289215,
 -0.027323827147483826,
 0.014530900865793228,
 0.05385929346084595,
 -0.012237491086125374,
 0.07152571529150009,
 0.017496207728981972,
 -0.015650730580091476,
 0.003997789230197668,
 -0.0028421259485185146,
 -0.021196123212575912,
 0.02621295675635338,
 0.059055302292108536,
 0.04550984874367714,
 -0.05568685382604599,
 0.03293193131685257,
 0.04052884876728058,
 0.005285592749714851,
 0.0030862486455589533,
 -0.029

In [9]:
len(query_result)

1024

In [9]:
from langchain_community.document_loaders import TextLoader

loader = TextLoader("speech.txt")
docs=loader.load()
docs

[Document(metadata={'source': 'speech.txt'}, page_content='The world must be made safe for democracy. Its peace must be planted upon the tested foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we shall freely make. We are but one of the champions of the rights of mankind. We shall be satisfied when those rights have been made as secure as the faith and the freedom of nations can make them.\n\nJust because we fight without rancor and without selfish object, seeking nothing for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident, conduct our operations as belligerents without passion and ourselves observe with proud punctilio the principles of right and of fair play we profess to be fighting for.\n\n…\n\nIt will be all the easier for us to conduct ourselves as belligerents in a high spirit of right and fairness be

In [11]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=200)
final_documents = text_splitter.split_documents(docs)
len(final_documents)

10

In [12]:
final_documents[0]

Document(metadata={'source': 'speech.txt'}, page_content='The world must be made safe for democracy. Its peace must be planted upon the tested foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we shall freely make. We are but one of the champions of the rights of mankind. We shall be satisfied when those rights have been made as secure as the faith and the freedom of nations can make them.')

In [14]:
embeddings.embed_query(final_documents[0].page_content)

[0.0298919677734375,
 -0.022735595703125,
 -0.006710052490234375,
 0.07861328125,
 0.01450347900390625,
 0.0109710693359375,
 -0.0246124267578125,
 0.021514892578125,
 0.034637451171875,
 0.032196044921875,
 0.019866943359375,
 0.13818359375,
 -0.032440185546875,
 -0.018035888671875,
 -0.006153106689453125,
 -0.0204925537109375,
 -0.03424072265625,
 0.016845703125,
 -0.059234619140625,
 0.0015954971313476562,
 -0.0247344970703125,
 -0.00676727294921875,
 -0.033111572265625,
 -0.0018749237060546875,
 0.0321044921875,
 0.02960205078125,
 -0.00844573974609375,
 0.0109710693359375,
 0.04058837890625,
 0.072509765625,
 -0.0011034011840820312,
 0.0185394287109375,
 0.036407470703125,
 -0.0029430389404296875,
 0.080322265625,
 0.01136016845703125,
 0.00806427001953125,
 0.0184173583984375,
 -0.01806640625,
 0.06549072265625,
 -0.00807952880859375,
 -0.024383544921875,
 0.0406494140625,
 0.0010995864868164062,
 0.062744140625,
 0.042999267578125,
 -0.010711669921875,
 -0.0010585784912109375,
 